src/MUSI6106Exec/MUSI6106Exec.cpp

```diff
@@ -1,131 +1,311 @@
+    int blockSize;
+    long long fileNumSample;
+    float** buffer;
+    CAudioFileIf::FileIoType_t ioType;
+    CAudioFileIf::FileSpec_t* fileSpec = nullptr;
+    CAudioFileIf* phAudioFile = nullptr;
+};
+
+/////////////////////////////////////////////////////////////////////////////////
+// main function
+int main(int argc, char* argv[])
+{
+    std::string             sInputFilePath,                 //!< file paths
+                            sIRFilePath,
+                            sOutputFilePath;
+
+    CFastConv* pCFastConv = new CFastConv();
 
-    ppfOutputAudio = new float* [stFileSpec.iNumChannels];
-    for (int i = 0; i < stFileSpec.iNumChannels; i++)
-        ppfOutputAudio[i] = new float[kBlockSize];
+    // command line args
+    ConvolverArgs_t args = parseArg(argc, argv);
+    sInputFilePath = args.inputPath;
+    sIRFilePath = args.IRPath;
+    sOutputFilePath = args.outputPath;
 
-    // Set parameters of vibrato
-    pCVibrato->setParam(CVibrato::kParamModFreqInHz, fModFrequencyInHz);
-    pCVibrato->setParam(CVibrato::kParamModWidthInS, fModWidthInSec);
+    ///////////////////////////////////////////////////////////////////////////
+    AudioFileWrapper inputAudio(sInputFilePath, args.blockSize, CAudioFileIf::kFileRead);
+    AudioFileWrapper impulseResponse(sIRFilePath, -1, CAudioFileIf::kFileRead);
+    AudioFileWrapper outputAudio(sOutputFilePath, args.blockSize, CAudioFileIf::kFileWrite, inputAudio.getFileSpec());
 
+    ////////////////////////////////////////////////////////////////////////////
+    impulseResponse.readData(impulseResponse.getNumSample());
+    pCFastConv->init(impulseResponse.getBuffer()[0], impulseResponse.getNumSample(), args.blockSize, args.convolver);
+    pCFastConv->setWetGain(args.wetGain);
+
+    int iNumFrames = args.blockSize;
+    long long processDuration = 0;
+
+    ////////////////////////////////////////////////////////////////////////////
     // processing
-    while (!phAudioFile->isEof())
+    while (!inputAudio.isEof())
     {
-        phAudioFile->readData(ppfInputAudio, iNumFrames);
-        pCVibrato->process(ppfInputAudio, ppfOutputAudio, iNumFrames);
-        phAudioOutputFile->writeData(ppfOutputAudio, iNumFrames);
+        inputAudio.readData(iNumFrames);
+
+        auto start = std::chrono::steady_clock::now();
+        pCFastConv->process(outputAudio.getBuffer()[0], inputAudio.getBuffer()[0], iNumFrames);
+        processDuration += std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::steady_clock::now() - start).count();
+
+        outputAudio.writeData(iNumFrames);
     }
-    phAudioFile->getFileSpec(stFileSpec);
 
+    // flush remaining
+    float* remain = new float[impulseResponse.getNumSample() - 1];
+    memset(remain, 0, sizeof(float) * (impulseResponse.getNumSample() - 1));
+    pCFastConv->flushBuffer(remain);
+    // and write remaining to output
+    // it's ugly but currently no more elegant way to work around this
```

Comment:
Is the issue that the tail length exceeds the size of `outputAudioFile`'s internal buffer? If so, perhaps an overload of `writeData` (that takes a user-provided buffer) could be handy.

Reply:
That's a very clever idea to keep it clean!

Reply:
Fixed in 04995c5937e11713dd3747ea65f37650d2eec19f.